### PR TITLE
Implemented https://github.com/HelloKitty/PokemonGoDesktop.API Providing All Proto Definitons

### PR DESCRIPTION
**Pull request by HelloKitty**
Hi, this is a pretty massive pull request.

It mostly changes internals and not so much the public Logic portion which most users interface with. It's been gutted to make use of my PokemonGoDesktop.API Nuget package that implement net35 compatible proto classes generated from POGOProto's project.

This is a major update that opens up major possibilities for extending the bot. Additionally it had an under-the-hood refactor to reduce most request methods to single lines in Client.cs utilizing things like: AwaitableOnResponseFor which may be user-friendly enough to expose to developers who build stuff on the logic side.

For popular example it would take someone only minutes to implement something like lured Pokemon catching or incense Pokemon encountering.

### CODE REVIEW

**Comment by HelloKitty** [comment]
I closed my previous pr. I really recommend taking the leap to adopting these new proto defs.

**Comment by AlginMaduro** [comment]
I would suggest moving merging this pull request and cleaning up the project... Was going through the code yesterday, and it was hurring my eyes.

File: PokemonGo.RocketAPI/Client.cs [hunk 1]
```diff
@@ -109,29 +109,52 @@ public Client(ISettings settings)
         public double CurrentAltitude { get; private set; }
 
         public async Task<CatchPokemonResponse> CatchPokemon(ulong encounterId, string spawnPointGuid, double pokemonLat,
-            double pokemonLng, MiscEnums.Item pokeball)
+            double pokemonLng, ItemId pokeball)
         {
-            var customRequest = new Request.Types.CatchPokemonRequest
+            CatchPokemonMessage catchPokemon = new CatchPokemonMessage()
```
Comment (AlginMaduro) [inline PokemonGo.RocketAPI/Client.cs]: Use var

File: PokemonGo.RocketAPI/Client.cs [hunk 2]
```diff
@@ -109,29 +109,52 @@ public Client(ISettings settings)
...
+                SpawnPointId = spawnPointGuid,
+                HitPokemon = true,
+                NormalizedReticleSize = 1.950,
+                SpinModifier = 1,
+                NormalizedHitPosition = 1
             };
 
-            var catchPokemonRequest = RequestBuilder.GetRequest(_unknownAuth, CurrentLat, CurrentLng, CurrentAltitude,
-                new Request.Types.Requests
+            return await AwaitableOnResponseFor<CatchPokemonMessage, CatchPokemonResponse>(catchPokemon, RequestType.CatchPokemon);
```
Comment (AlginMaduro) [inline PokemonGo.RocketAPI/Client.cs]: Howcome the current location dissapeared?

**Comment by HelloKitty** [comment]
Hang on, found an issue with transfering. Need to look into what broke it.

File: PokemonGo.RocketAPI/Client.cs (hunk 2 above)
Comment (HelloKitty) [inline PokemonGo.RocketAPI/Client.cs]: Because AwaitableOnResponseFor handles that.

File: PokemonGo.RocketAPI/Client.cs (hunk 1 above)
Comment (HelloKitty) [inline PokemonGo.RocketAPI/Client.cs]: Same IL generated. Not important =P

File: PokemonGo.RocketAPI/Client.cs [hunk 3]
```diff
@@ -165,171 +188,103 @@ public Client(ISettings settings)
...
-            var customRequest = new Request.Types.EncounterRequest
+            EncounterMessage encounterPokemonMessage = new EncounterMessage()
             {
                 EncounterId = encounterId,
-                SpawnpointId = spawnPointGuid,
-                PlayerLatDegrees = Utils.FloatAsUlong(CurrentLat),
-                PlayerLngDegrees = Utils.FloatAsUlong(CurrentLng)
+                SpawnPointId = spawnPointGuid,
+                PlayerLatitude = CurrentLat,
+                PlayerLongitude = CurrentLng,
```
Comment (AlginMaduro) [inline PokemonGo.RocketAPI/Client.cs]: Use var and remove comma at the last param

File: PokemonGo.RocketAPI/Client.cs [hunk 4]
```diff
@@ -165,171 +188,103 @@ public Client(ISettings settings)
...
             {
-                CellIds =
-                    ByteString.CopyFrom(
-                        ProtoHelper.EncodeUlongList(S2Helper.GetNearbyCellIds(CurrentLng,
-                            CurrentLat))),
-                Latitude = Utils.FloatAsUlong(CurrentLat),
-                Longitude = Utils.FloatAsUlong(CurrentLng),
-                Unknown14 = ByteString.CopyFromUtf8("\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")
+                Latitude = CurrentLat,
+                Longitude = CurrentLng,
```
Comment (AlginMaduro) [inline PokemonGo.RocketAPI/Client.cs]: Use var, remove comma at last param. Also rve the (), u are already instancing the object with parameters. Do this for all object instancing in the class

Comment (HelloKitty) [inline PokemonGo.RocketAPI/Client.cs]: Var is pointless. Provides not benefit in this context.

Your second criticism makes no sense. The diff linked  showing no instantiation.

File: PokemonGo.RocketAPI/Client.cs (hunk 3 above)
Comment (HelloKitty) [inline PokemonGo.RocketAPI/Client.cs]: Comma can stay. Same IL generated.

**Comment by HelloKitty** [comment]
@AlginMaduro There is a point to making a pull request. This pull request isn't about prettying anything up. It's about unlocking every .proto def which this project was lacking.

File: PokemonGo.RocketAPI/Client.cs (hunk 1 above)
Comment (DaniilSokolyuk) [inline PokemonGo.RocketAPI/Client.cs]: Please read https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

> We only use var when it's obvious what the variable type is (i.e. var stream = new FileStream(...) not var stream = OpenStandardInput()).

Comment (HelloKitty) [inline PokemonGo.RocketAPI/Client.cs]: Those guidelines are violated throughout the codebase.

Plus, it doesn't explictly say to use var in those cases. Just that they only use var in those situations. Seems more of a standard of when not to use var than when to use var.

File: PokemonGo.RocketAPI.Console/PokemonGo.RocketAPI.Console.csproj [hunk 5]
```diff
@@ -32,8 +32,24 @@
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Google.Protobuf, Version=3.0.0.0, Culture=neutral, PublicKeyToken=a7d26565bac4d604, processorArchitecture=MSIL">
-      <HintPath>..\packages\Google.Protobuf.3.0.0-beta3\lib\dotnet\Google.Protobuf.dll</HintPath>
+    <Reference Include="Easyception, Version=2.0.0.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (AlginMaduro) [inline PokemonGo.RocketAPI.Console/PokemonGo.RocketAPI.Console.csproj]: I see a bunch of references to easyception, but could not find any refrences to it... Maybe im blind, but could u remove it if its not used?

**Comment by HelloKitty** [comment]
Fixed the issue with transferring. Was an issue with changed proto def and my assumption of its value.

It's now fully working as far as I can tell.

**Comment by AlginMaduro** [comment]
I know its not about making it pretty, sorry for this.. Used to it from work :D
Other then that its looks good.
Just a bit annoyed that the project is such a mess :p

File: PokemonGo.RocketAPI.Console/PokemonGo.RocketAPI.Console.csproj (hunk 5 above)
Comment (HelloKitty) [inline PokemonGo.RocketAPI.Console/PokemonGo.RocketAPI.Console.csproj]: No, it's a dependency of one of the newly introduced dependencies.

Also, at this point it seems like you're trolling this PR.

**Comment by HelloKitty** [comment]
@AlginMaduro Yes, it is definitely a mess. I tried to client up Client.cs a tiny bit by consolidating envelope building in that one generic async method.

I didn't cause the mess though so I tread lightly. I just wanted to introduce every .proto from POGOProtos. This PR achieves that and it's the only true point to the PR.

**Comment by AlginMaduro** [comment]
@HelloKitty indeed, and u did a great job of it :D

**Comment by HelloKitty** [comment]
@AlginMaduro Thanks, I can't wait to see it in use too! Literally all these are in the project now https://github.com/HelloKitty/PokemonGoDesktop.API/tree/master/src/PokemonGoDesktop.API.Proto it'll be cool to see what features people build with them.

File: PokemonGo.RocketAPI.Console/PokemonGo.RocketAPI.Console.csproj (hunk 5 above)
Comment (AlginMaduro) [inline PokemonGo.RocketAPI.Console/PokemonGo.RocketAPI.Console.csproj]: not trolling.. just reviewing :P

**Comment by pdtan** [comment]
how to fix transfer?

**Comment by HelloKitty** [comment]
Transfer isn't broken in this PR.

This code isn't apart of the main repo right now so asking about transferring with Necro's bot currently in here will not help @pdtan .

**Comment by pdtan** [comment]
I used your API, and you said 'Hang on, found an issue with transfering.', then I test, transfer not working

**Comment by pdtan** [comment]
Ah, fixed. Thanks String.IsNullOrEmpty()

**Comment by HelloKitty** [comment]
@pdtan Oh very sorry, I didn't know that you were talking about this PR. I thought maybe someone from #help was asking a question lol. My apologies.

Yea, that FortId thing caused me a headache. I couldn't find out what was causing Pokemon to stop transfering until I traced through the diff and realized that maybe the new proto def that had a string FortId was empty instead of null.

**Comment by pdtan** [comment]
You saved me =))
DeployedFortId always returns 0 when using Rocket API, I discovered its tag is 66 (instead of 64) but don't know how to fix.

**Comment by JoshYuen** [comment]
Lets get this thing in! this is a major improvement.

**Comment by HelloKitty** [comment]
It came to my attention that Sunrise did something similar yesterday to Ferox https://github.com/FeroxRev/Pokemon-Go-Rocket-API/pull/21 . I wasn't aware of it.

**Comment by navidemad** [comment]
evolve

I don't know if it is related to your PR, but with a fresh folder from your repository, i got this error msg.

**Comment by HelloKitty** [comment]
@navidemad It's not related unless you're currently using my fork which is not available yet in this repo.

**Comment by theplaymate** [comment]
@HelloKitty is it possible to merge your PR with my forked repo of this somehow? I sooo want to test this

**Comment by HelloKitty** [comment]
@theplaymate  I'd imagine there would be merge conflicts if you tried. You can check out my fork if you want.

**Comment by navidemad** [comment]
I fork your repository @HelloKitty , with the following settings and no evolution worked.
settings

**Comment by theplaymate** [comment]
@HelloKitty the forked one is the "ProtoUpdate" branch right? Because the master branch is just 1 commit ahead

**Comment by HelloKitty** [comment]
@theplaymate Yes, ProtoUpdate.

@navidemad You forked my repo? Why? Are you saying Evolution is working in the current repo but is not working in my fork? Also, are you using the right branch in my fork?

**Comment by navidemad** [comment]
@HelloKitty I forked your repo with [the branch ProtoUpdate](https://github.com/HelloKitty/NecroBot/tree/ProtoUpdate), and yes Evolution is working in the current repo but not in your repo. I just tried it right now.

**Comment by HelloKitty** [comment]
@navidemad Thank you for finding that. I'll look and see what could of caused that now.

**Comment by HelloKitty** [comment]
@navidemad Thanks for testing this out, was one of the few things I didn't test. It was a typical copy-paste error. I'd like to have condensed most of Client.cs into some sort of magical method without all this code duplication but couldn't.

Either way, I was sending the Encounter request ID to the server instead of the Evolve. A new push in a moment will fix it.

**Comment by navidemad** [comment]
Nice commit, it works now good job :+1:

**Comment by navidemad** [comment]
Sorry to bother you again, but if you can fix this one too :+1: 

exception

**Comment by HelloKitty** [comment]
@navidemad I've run this quite abit and never encountered that. You sure you didn't change anything? Very strange. I'll keep looking but I can't find anything so far.

**Comment by HelloKitty** [comment]
@navidemad Do you happen to have any Pokemon deployed at the Gym perchance?

**Comment by navidemad** [comment]
I did not change anything and i have not pokemon deployed at any Gym right now.

**Comment by HelloKitty** [comment]
@navidemad I believe this most recent commit d7f0b13 will fix it. I'm not really sure I changed that though >.>. I can't test it since I did not encounter the issue sadly and we have no unit tests.

**Comment by navidemad** [comment]
@HelloKitty alright, i update it and gonna test it. I will keep you informed tomorrow morning (US localtime)

**Comment by theplaymate** [comment]
@HelloKitty I tried creating a method for using egg incubator and it turned out the "ItemId" property of "UseItemEggIncubatorMessage" expects a string instead of and "ItemId" enum. Will it work fine if i ToString the enum and pass it in?

public async Task<UseItemEggIncubatorResponse> UseItemEggIncubator(ItemId itemId, ulong pokemonId)
        {
            UseItemEggIncubatorMessage useEggIncubatorMessage = new UseItemEggIncubatorMessage
            {
                ItemId = itemId,
                PokemonId = pokemonId
            };
            return
                await
                    AwaitableOnResponseFor<UseItemEggIncubatorMessage, UseItemEggIncubatorResponse>(
                        useEggIncubatorMessage, RequestType.UseItemEggIncubator);
        }

**Comment by HelloKitty** [comment]
@navidemad You don't encounter it anymore? If that commit fixed it then it was a fault in Necro's current repo too. Let me know, I need to sleep too. It's 5am here lol.

**Comment by navidemad** [comment]
@HelloKitty I will tell you if it occured again in the next couple hours.

**Comment by HelloKitty** [comment]
@theplaymate It's probably an ItemId like you said. Encode the int value of the Enum you want to send using UTF8 temporarily. I think that will work for now.

This is more directed towards POGOProtos though since we're downstream of them right now. Raise an issue over there, I can't support that with this pull request.

**Comment by theplaymate** [comment]
@HelloKitty alright, i'll do that now. Thanks

**Comment by theplaymate** [comment]
@HelloKitty sorry to interrupt you again but i found 2 properties inside an egg incubator. "ItemId" and "Id", where "Id" is a string identifier. Maybe this is the Id that the request is expecting. Not sure though, but will test with both.

**Comment by HelloKitty** [comment]
I only see the two fields Id and PokemonId. I think ItemId in the .poro should be the ItemId enum but it's string so you'll have to store the int that .proto would otherwise use as UTF8 I think.

The PokemonId is probably the egg's GUID. It's definitely an incorrect proto but it should be fixed on POGOProto's side.

**Comment by theplaymate** [comment]
@HelloKitty its the "Id" property, just tested. Because "ItemId" only lets you know the type of Item, not which item of that type

**Comment by theplaymate** [comment]
@HelloKitty this is what I found in EggIncubator and i think "Id" is probably the right identifier for the egg
capture

**Comment by HelloKitty** [comment]
Hmmm, I didn't think each item had a unique id like that. Other UseItem messages don't define an ID like that such as this one https://github.com/AeonLucid/POGOProtos/blob/master/src/POGOProtos/Networking/Requests/Messages/UseItemPotionMessage.proto

But I suppose incubators could special since they interface with eggs and probably register themselves somehow on the serverside for incubation unlike a potion that simply heals you.

**Comment by theplaymate** [comment]
Exactly thats what i was thinking. Each egg is incubated in a specific Incubator thats why it has a unique identifier,

**Comment by HelloKitty** [comment]
@theplaymate Ahhh, guess we learned never to doubt POGOProtos. They really are on the ball with their protobuf defs. Hope you're finding it easier to add new features thanks to this PR.

I have to sleep now so hopefully nothing else arises. Hopefully they merge this. If they don't I'll still maintain the Nuget packages it's based on if you want to continue to use it.

**Comment by theplaymate** [comment]
@HelloKitty your repo is helping a lot, easy to add more features. Will continue to use yours. Keep up the good work. And goodnight!

**Comment by MonsterMMORPG** [comment]
if you have a working incubate eggs feature i will migrate to your project 
do you have right now?

**Comment by pdtan** [comment]
Very easy to add eggs into Incubators with your code. Thanks

**Comment by Degix** [comment]
Finally!... I can code a routine for gyms :dancer: 
Cant wait anymore ...

**Comment by cloudsstars** [comment]
kinda new to this how do i add this to mine i been looking for one that will hatch my eggs or can someone post the zip for me or tell me how to add please and thank you .this looks like an awesome project

**Comment by AlginMaduro** [comment]
@NecronomiconCoding i think its becomming time to merge thisone...

**Comment by Ilses** [comment]
Since this seems to be getting merged soon, I'll submit a new PR for catching lured pokemon based on this PR

**Comment by cloudsstars** [comment]
@pdtan  how did you add egg Incubators would be much help

**Comment by HelloKitty** [comment]
I will be closing this pull request since they've something similar implemented in Ferox.

**Pull request closed without merging**